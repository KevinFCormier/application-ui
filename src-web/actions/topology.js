/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/

import lodash from 'lodash'

import * as Actions from './index'
import { RESOURCE_TYPES } from '../../lib/shared/constants'
import apolloClient from '../../lib/client/apollo-client'

export const requestResource = (resourceType, fetchFilters, reloading) => ({
  type: Actions.RESOURCE_REQUEST,
  status: Actions.REQUEST_STATUS.IN_PROGRESS,
  resourceType,
  fetchFilters,
  reloading
})

export const receiveResourceError = (err, resourceType) => ({
  type: Actions.RESOURCE_RECEIVE_FAILURE,
  status: Actions.REQUEST_STATUS.ERROR,
  err,
  resourceType
})

export const receiveTopologySuccess = (
  response,
  resourceType,
  fetchFilters,
  willLoadDetails
) => ({
  type: Actions.RESOURCE_RECEIVE_SUCCESS,
  status: Actions.REQUEST_STATUS.DONE,
  nodes: response.resources || [],
  links: response.relationships || [],
  filters: {
    clusters: response.clusters,
    labels: response.labels,
    namespaces: response.namespaces,
    types: response.resourceTypes
  },
  resourceType,
  fetchFilters,
  willLoadDetails
})

export const requestResourceDetails = (
  resourceType,
  fetchFilters,
  reloading
) => ({
  type: Actions.RESOURCE_DETAILS_REQUEST,
  status: Actions.REQUEST_STATUS.IN_PROGRESS,
  resourceType,
  fetchFilters,
  reloading
})

export const receiveTopologyDetailsSuccess = (
  response,
  resourceType,
  fetchFilters
) => ({
  type: Actions.RESOURCE_DETAILS_RECEIVE_SUCCESS,
  status: Actions.REQUEST_STATUS.DONE,
  pods: response.pods || [],
  resourceType,
  fetchFilters
})

export const fetchTopology = (vars, fetchFilters, reloading) => {
  const resourceType = RESOURCE_TYPES.HCM_TOPOLOGY
  return dispatch => {
    dispatch(requestResource(resourceType, fetchFilters, reloading))
    apolloClient
      .getResource(resourceType, vars)
      .then(response => {
        if (response.errors) {
          dispatch(receiveResourceError(response.errors[0], resourceType))
        } else {
          // return topology
          const topology = {
            clusters: lodash.cloneDeep(response.data.clusters),
            labels: lodash.cloneDeep(response.data.labels),
            namespaces: lodash.cloneDeep(response.data.namespaces),
            resourceTypes: lodash.cloneDeep(response.data.resourceTypes),
            resources: lodash.cloneDeep(response.data.topology.resources),
            relationships: lodash.cloneDeep(
              response.data.topology.relationships
            )
          }
          dispatch(
            receiveTopologySuccess(topology, resourceType, fetchFilters, false)
          )
        }
      })
      .catch(err => dispatch(receiveResourceError(err, resourceType)))
  }
}

export const restoreSavedTopologyFilters = (namespace, name) => ({
  type: Actions.TOPOLOGY_RESTORE_SAVED_FILTERS,
  namespace,
  name
})

export const updateTopologyFilters = (
  filterType,
  filters,
  namespace,
  name
) => ({
  type: Actions.TOPOLOGY_FILTERS_UPDATE,
  filterType,
  filters,
  namespace,
  name
})

const receiveFiltersError = err => ({
  type: Actions.TOPOLOGY_FILTERS_RECEIVE_ERROR,
  err
})

export const fetchTopologyFilters = () => {
  return dispatch => {
    dispatch({
      type: Actions.TOPOLOGY_FILTERS_REQUEST
    })
    return apolloClient
      .getTopologyFilters()
      .then(response => {
        if (response.errors) {
          return dispatch(receiveFiltersError(response.errors[0]))
        }
        return dispatch({
          type: Actions.TOPOLOGY_FILTERS_RECEIVE_SUCCESS,
          clusters: lodash.cloneDeep(response.data.clusters),
          labels: lodash.cloneDeep(response.data.labels),
          namespaces: lodash.cloneDeep(response.data.namespaces),
          types: lodash.cloneDeep(response.data.resourceTypes)
        })
      })
      .catch(err => dispatch(receiveFiltersError(err)))
  }
}
