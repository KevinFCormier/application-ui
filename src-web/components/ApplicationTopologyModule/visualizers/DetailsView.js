/*******************************************************************************
 * Licensed Materials - Property of IBM
 * 5737-E67
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 *
 * US Government Users Restricted Rights - Use, duplication or disclosure
 * restricted by GSA ADP Schedule Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import { Icon } from 'carbon-components-react'
import { DetailsViewDecorator } from './DetailsViewDecorator'
import { Scrollbars } from 'react-custom-scrollbars'
import resources from '../../../../lib/shared/resources'
import msgs from '../../../../nls/platform.properties'

resources(() => {
  require('../scss/topology-details.scss')
})

class DetailsView extends React.Component {
  constructor(props) {
    super(props)
  }

  handleClick(value) {
    this.fetchLogs(value)
  }

  handleKeyPress(value, e) {
    if (e.key === 'Enter') {
      this.fetchLogs(value)
    }
  }

  fetchLogs(value) {
    const { fetchLogs, onClose } = this.props
    const { resourceType, data } = value
    fetchLogs(resourceType, data)
    onClose()
  }

  render() {
    const {
      context,
      onClose,
      getLayoutNodes,
      getViewContainer,
      staticResourceData,
      selectedNodeId
    } = this.props
    const { typeToShapeMap, getNodeDetails } = staticResourceData
    const currentNode =
      getLayoutNodes().find(n => n.uid === selectedNodeId) || {}
    const { layout = {} } = currentNode
    const resourceType = layout.type || currentNode.type
    const { shape = 'circle', className = 'default' } =
      typeToShapeMap[resourceType] || {}
    const details = getNodeDetails(currentNode, context)
    const name = currentNode.name
    const height = getViewContainer().getBoundingClientRect().height
    const scrollHeight = height * 0.75
    return (
      <section className={`topologyDetails ${className}`}>
        <h3 className="detailsHeader">
          <DetailsViewDecorator shape={shape} className={className} />
          <span className="titleText">{name}</span>
          <Icon
            className="closeIcon"
            description={msgs.get('topology.details.close', context.locale)}
            name="icon--close"
            onClick={onClose}
          />
        </h3>
        <hr />
        <Scrollbars
          style={{ width: 310, height: scrollHeight }}
          renderView={this.renderView}
          renderThumbVertical={this.renderThumbVertical}
          className="details-view-container"
        >
          {details.map(({ type, labelKey, value }) => {
            let handleClick, handleKeyPress
            if (type === 'logs') {
              handleClick = this.handleClick.bind(this, value)
              handleKeyPress = this.handleKeyPress.bind(this, value)
            } else {
              handleClick = handleKeyPress = null
            }
            return type === 'spacer' ? (
              <div className="sectionContent" key={Math.random()}>
                <div className="spacer" />
              </div>
            ) : type !== 'logs' ? (
              <div className="sectionContent" key={Math.random()}>
                {labelKey && (
                  <span className="label">
                    {msgs.get(labelKey, context.locale)}:{' '}
                  </span>
                )}
                <span className="value">{value}</span>
              </div>
            ) : (
              <div className="sectionContent" key={Math.random()}>
                <div
                  className="logs"
                  tabIndex="0"
                  role={'button'}
                  onClick={handleClick}
                  onKeyPress={handleKeyPress}
                >
                  {msgs.get('topology.view.logs', context.locale)}
                </div>
              </div>
            )
          })}
        </Scrollbars>
      </section>
    )
  }

  renderView({ style, ...props }) {
    style.marginBottom = -17
    return <div {...props} style={{ ...style }} />
  }

  renderThumbVertical({ style, ...props }) {
    const finalStyle = {
      ...style,
      cursor: 'pointer',
      borderRadius: 'inherit',
      backgroundColor: 'rgba(0,0,0,.2)'
    }
    return (
      <div className={'details-view-scrollbar'} style={finalStyle} {...props} />
    )
  }
}

DetailsView.propTypes = {
  context: PropTypes.object,
  fetchLogs: PropTypes.func,
  getLayoutNodes: PropTypes.func,
  getViewContainer: PropTypes.func,
  onClose: PropTypes.func,
  selectedNodeId: PropTypes.string,
  staticResourceData: PropTypes.object
}

export default DetailsView
