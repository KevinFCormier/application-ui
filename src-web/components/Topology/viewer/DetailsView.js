/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018, 2019. All Rights Reserved.
 * Copyright (c) 2020 Red Hat, Inc.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import _ from 'lodash'
import R from 'ramda'
import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import { Scrollbars } from 'react-custom-scrollbars'
import {
  Icon,
  NumberInput,
  MultiSelect,
  Button
} from 'carbon-components-react'
import jsYaml from 'js-yaml'
import '../../../../graphics/diagramShapes.svg'
import msgs from '../../../../nls/platform.properties'

const DetailsViewDecorator = ({ shape, className }) => {
  return (
    <svg width="48px" height="48px" viewBox="0 0 48 48">
      <use
        href={`#diagramShapes_${shape}`}
        className={`${className} detailsIcon`}
      />
    </svg>
  )
}

DetailsViewDecorator.propTypes = {
  className: PropTypes.string,
  shape: PropTypes.string
}

class DetailsView extends React.Component {
  constructor(props) {
    super(props)
  }

  handleClick(value) {
    this.processActionLink(value)
  }

  handleKeyPress(value, e) {
    if (e.key === 'Enter') {
      this.processActionLink(value)
    }
  }

  processActionLink(value) {
    const { processActionLink } = this.props
    const { data } = value
    processActionLink(data)
  }

  render() {
    const {
      locale,
      onClose,
      getLayoutNodes,
      getViewContainer,
      staticResourceData,
      selectedNodeId
    } = this.props
    const { typeToShapeMap, getNodeDetails } = staticResourceData
    const currentNode =
      getLayoutNodes().find(n => n.uid === selectedNodeId) || {}
    const { layout = {} } = currentNode
    const resourceType = layout.type || currentNode.type
    const { shape = 'other', className = 'default' } =
      typeToShapeMap[resourceType] || {}
    const details = getNodeDetails(currentNode, locale)
    const name = currentNode.name
    const height = getViewContainer().getBoundingClientRect().height
    const scrollHeight = height * 0.75
    return (
      <section className="topologyDetails">
        <div className="detailsHeader">
          <DetailsViewDecorator shape={shape} className={className} />
          <div>
            <div className="sectionContent">
              <span className="label">{_.capitalize(resourceType)}</span>
            </div>
            <div className="sectionContent">
              <span className="titleNameText">{name}</span>
            </div>
          </div>
          <Icon
            className="closeIcon"
            description={msgs.get('topology.details.close', locale)}
            name="icon--close"
            onClick={onClose}
          />
        </div>
        <hr />
        <Scrollbars
          style={{ height: scrollHeight }}
          renderView={this.renderView}
          renderThumbVertical={this.renderThumbVertical}
          className="details-view-container"
        >
          {details.map(detail => this.renderDetail(detail, locale))}
        </Scrollbars>
      </section>
    )
  }

  renderDetail(detail, locale) {
    switch (detail.type) {
    case 'spacer':
      return this.renderSpacer()
    case 'link':
      return this.renderLink(detail, true)
    case 'number':
      return this.renderNumber(detail, locale)
    case 'selector':
      return this.renderSelector(detail, locale)
    case 'submit':
      return this.renderSubmit(detail, locale)
    case 'snippet':
      return this.renderSnippet(detail, locale)
    default:
      return this.renderLabel(detail, locale)
    }
  }

  renderLabel({ labelKey, labelValue, value, indent, status }, locale) {
    let label = labelValue
    if (labelKey) {
      label = labelValue
        ? msgs.get(labelKey, [labelValue], locale)
        : msgs.get(labelKey, locale)
    }
    label = value ? `${label}:` : label

    const mainSectionClasses = classNames({
      sectionContent: true,
      borderLeft: value !== undefined ? true : false
    })
    const labelClass = classNames({
      label: true,
      sectionLabel: value ? true : false
    })

    const statusIcon = status ? `icon--${status}` : undefined
    return (
      <div className={mainSectionClasses} key={Math.random()}>
        {(labelKey || labelValue) && statusIcon ? (
          <span className="label sectionLabel">
            <Icon
              name={statusIcon}
              fill="#6089bf"
              description=""
              className="label-icon"
            />
            {label}
          </span>
        ) : (
          <span className={labelClass}>{label} </span>
        )}
        {indent && <span className="indent" />}
        <span className="value">{value}</span>
      </div>
    )
  }

  renderSnippet({ value }) {
    if (value) {
      const yaml = jsYaml.safeDump(value).split('\n')
      return (
        <div className="sectionContent snippet">
          {yaml.map(line => {
            return <code key={Math.random()}>{line}</code>
          })}
        </div>
      )
    }
    return null
  }

  renderLink({ value, indent }) {
    const handleClick = this.handleClick.bind(this, value)
    const handleKeyPress = this.handleKeyPress.bind(this, value)
    const showLaunchOutIcon = !R.pathOr(false, ['data', 'specs', 'isDesign'])(
      value
    ) //if not show yaml

    let iconName = 'icon--arrow--right'
    if (_.get(value, 'label', '').startsWith('http')) {
      iconName = 'icon--launch'
    }
    const mainSectionClasses = classNames({
      sectionContent: true,
      borderLeft: indent ? true : false
    })

    const linkLabelClasses = classNames({
      link: true,
      sectionLabel: indent ? true : false
    })

    return (
      <div className={mainSectionClasses} key={Math.random()}>
        <span
          className={linkLabelClasses}
          id="linkForNodeAction"
          tabIndex="0"
          role={'button'}
          onClick={handleClick}
          onKeyPress={handleKeyPress}
        >
          {value.label}
          {showLaunchOutIcon && (
            <Icon
              name={iconName}
              fill="#6089bf"
              description=""
              className="label-icon"
            />
          )}
        </span>
      </div>
    )
  }

  renderNumber({ labelKey, value }, locale) {
    return (
      <div className="sectionContent form" key={Math.random()}>
        {labelKey && (
          <span className="label">{msgs.get(labelKey, locale)}: </span>
        )}
        <NumberInput
          id={labelKey}
          label={''}
          value={value}
          min={1}
          max={10}
          step={1}
          onChange={this.enableSubmitBtn.bind(this)}
        />
      </div>
    )
  }

  renderSelector({ labelKey, value, other }, locale) {
    return (
      <div className="sectionContent form" key={Math.random()}>
        {labelKey && (
          <span className="label">{msgs.get(labelKey, locale)}: </span>
        )}
        <MultiSelect.Filterable
          items={other}
          initialSelectedItems={value}
          placeholder={value.join(', ')}
          itemToString={item => item}
          onChange={this.enableSubmitBtn.bind(this)}
        />
      </div>
    )
  }

  setSubmitBtn = ref => {
    this.submitBtn = ref
  };
  renderSubmit() {
    return (
      <div className="sectionContent submit" key={Math.random()}>
        <Button
          small
          id="edit-button"
          disabled
          ref={this.setSubmitBtn}
          onClick={this.onSubmit}
        >
          {msgs.get('modal.button.submit', this.props.locale)}
        </Button>
      </div>
    )
  }

  enableSubmitBtn() {
    this.submitBtn.disabled = false
  }

  onSubmit() {
    this.props.onClose()
  }

  renderSpacer() {
    return (
      <div className="sectionContent" key={Math.random()}>
        <div className="spacer" />
      </div>
    )
  }

  renderView({ style, ...props }) {
    style.marginBottom = -17
    return <div {...props} style={{ ...style }} />
  }

  renderThumbVertical({ style, ...props }) {
    const finalStyle = {
      ...style,
      cursor: 'pointer',
      borderRadius: 'inherit',
      backgroundColor: 'rgba(0,0,0,.2)'
    }
    return (
      <div className={'details-view-scrollbar'} style={finalStyle} {...props} />
    )
  }
}

DetailsView.propTypes = {
  getLayoutNodes: PropTypes.func,
  getViewContainer: PropTypes.func,
  locale: PropTypes.string,
  onClose: PropTypes.func,
  processActionLink: PropTypes.func,
  selectedNodeId: PropTypes.string,
  staticResourceData: PropTypes.object
}

export default DetailsView
