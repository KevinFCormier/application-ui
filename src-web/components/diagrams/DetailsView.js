/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import { Icon } from 'carbon-components-react'
import { DetailsViewDecorator } from './DetailsViewDecorator'
import msgs from '../../../nls/platform.properties'

class DetailsView extends React.Component {

  constructor(props) {
    super(props)
  }

  handleClick(value) {
    this.fetchLogs(value)
  }

  handleKeyPress(value, e) {
    if ( e.key === 'Enter') {
      this.fetchLogs(value)
    }
  }

  fetchLogs(value) {
    const { fetchLogs, onClose } = this.props
    const { resourceType, data } = value
    fetchLogs(resourceType, data)
    onClose()
  }

  render() {
    const { context, onClose, getLayoutNodes, staticResourceData, selectedNodeId} = this.props
    const {typeToShapeMap, getNodeDetails} = staticResourceData
    const currentNode = getLayoutNodes().find((n) => n.uid === selectedNodeId) || {}
    const { layout={} } = currentNode
    const resourceType = layout.type || currentNode.type
    const {shape='circle', className='default'} =  typeToShapeMap[resourceType]||{}
    const details = getNodeDetails(currentNode, context)
    const name = currentNode.name
    return (
      <section className={`topologyDetails ${className}`}>
        <h3 className='detailsHeader'>
          <DetailsViewDecorator
            shape={shape}
            className={className}
          />
          <span className='titleText'>
            {name}
          </span>
          <Icon
            className='closeIcon'
            description={msgs.get('topology.details.close', context.locale)}
            name="icon--close"
            onClick={onClose}
          />
        </h3>
        <hr />
        {details.map(({type, labelKey, value}) =>
        {
          let handleClick, handleKeyPress
          if (type==='logs') {
            handleClick = this.handleClick.bind(this, value)
            handleKeyPress = this.handleKeyPress.bind(this, value)
          } else {
            handleClick = handleKeyPress = null
          }
          return (type==='spacer' ?
            <div className='sectionContent' key={Math.random()}>
              <div className='spacer'></div>
            </div> : (type!=='logs' ?
              <div className='sectionContent' key={Math.random()}>
                {labelKey && <span className='label'>{msgs.get(labelKey, context.locale)}: </span>}
                <span className='value'>{value}</span>
              </div> :
              <div className='sectionContent' key={Math.random()}>
                <div className='logs' tabIndex='0' role={'button'}
                  onClick={handleClick} onKeyPress={handleKeyPress}>
                  {msgs.get('topology.view.logs', context.locale)}
                </div>
              </div>))}
        )}
      </section>)
  }
}

DetailsView.propTypes = {
  context: PropTypes.object,
  fetchLogs: PropTypes.func,
  getLayoutNodes: PropTypes.func,
  onClose: PropTypes.func,
  selectedNodeId: PropTypes.string,
  staticResourceData: PropTypes.object,
}

export default DetailsView

