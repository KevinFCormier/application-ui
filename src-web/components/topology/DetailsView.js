/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import PropTypes from 'prop-types'
import msgs from '../../../nls/platform.properties'

const DetailsView = ({ context, details = [], title = '', resourceType = 'default' , /* status */ }) => (
  <section className={`topologyDetails ${resourceType}`}>
    <h3>{title}</h3>
    <hr />
    {details.map((d) =>
      <div className='sectionContent' key={d.labelKey}>
        <span className='label'>{msgs.get(d.labelKey, context.locale)}: </span>
        <span className='value'>{d.value}</span>
      </div>
    )}
  </section>
)

DetailsView.propTypes = {
  context: PropTypes.object,
  details: PropTypes.arrayOf(PropTypes.shape({
    type: PropTypes.oneOf(['label']).isRequired,
    labelKey: PropTypes.string,
    value: PropTypes.string,
  })),
  resourceType: PropTypes.string,
  title: PropTypes.string
}

export { DetailsView }
