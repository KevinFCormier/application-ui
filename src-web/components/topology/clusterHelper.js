/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import * as d3 from 'd3'
import { CLUSTER_MARGIN, CLUSTER_MARGIN_TOP } from './constants'


export default class ClusterHelper {
  /**
   * Helper class to be used by TopologyDiagram.
   *
   * Contains functions to draw a cluster in the diagram.
   */
  constructor(height, width, clusters){
    this.clusters = clusters
    this.height = height
    this.width = width
    this.svg = d3.select('svg.topologyDiagram')
  }

  removeOldClustersFromDiagram = () => {
    // TODO: We could preserve the existing clusters using
    //      `data(this.clusters, (n) => n.id).exit()`
    //      But we have to resize the remaining clusters. A
    //      simpler approach is to remove all clusters and redraw.
    this.svg.select('g.clusters')
      .selectAll('g.cluster')
      .remove()
  }

  addClustersToDiagram = () => {
    const clusterQuantity = this.clusters ? this.clusters.length : 0

    const clusters = this.svg.select('g.clusters')
      .selectAll('g.cluster')
      .data(this.clusters, (n) => n.id)
      .enter().append('g')
      .attr('id', (n) => `cluster-${n.name}`)
      .attr('class', 'cluster')

    clusters.append('rect')
      .attr('class', 'clusterArea')
      .attr('x', CLUSTER_MARGIN).attr('y', CLUSTER_MARGIN_TOP)
      .attr('height', this.height - CLUSTER_MARGIN_TOP * 2)
      .attr('width', `${this.width / clusterQuantity - CLUSTER_MARGIN * 2}`)
      .attr('transform', (n) => `translate(${this.width * n.index / clusterQuantity}, 0)`)
    clusters.append('text').attr('class','clusterTitle').text((n) =>`Cluster: ${n.name}`)
      .attr('transform', (n) => `translate(${this.width * n.index / clusterQuantity + CLUSTER_MARGIN}, 45)`)

  }
}
