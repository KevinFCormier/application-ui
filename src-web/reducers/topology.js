/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/

import lodash from 'lodash'
import * as Actions from '../actions'
import { RESOURCE_TYPES, HCM_TOPOLOGY_FILTER_COOKIE } from '../../lib/shared/constants'

const initialState = {
  availableFilters: {
    clusters: [],
    labels: [],
    namespaces: [],
    types: [],
  },
  activeFilters: {
    namespace: [],
    type: [{ label: 'deployment'}] // Sets the default filters
  },
  otherTypeFilters: [],
  links: [],
  nodes: [],
  status: Actions.REQUEST_STATUS.INCEPTION,
}

//default k8 type filters
export const defaultTypeFilters = new Set(['internet', 'host', 'service', 'deployment',
  'daemonset', 'statefulset', 'pod', 'container'])


export const topology = (state = initialState, action) => {
  if (action.resourceType && action.resourceType.name === RESOURCE_TYPES.HCM_TOPOLOGY.name){
    switch (action.type) {
    case Actions.RESOURCE_REQUEST: {
      return {...state, status: Actions.REQUEST_STATUS.IN_PROGRESS}
    }
    case Actions.RESOURCE_RECEIVE_SUCCESS: {

      // ignore topologies that were fetched with a different set of active filters
      if (lodash.isEqual(action.fetchFilters, state.activeFilters)) {
        return { ...state,
          status: Actions.REQUEST_STATUS.DONE,
          nodes: action.nodes,
          links: action.links,
          fetchFilters: action.fetchFilters
        }
      } else {
        return { ...state }
      }
    }
    case Actions.RESOURCE_RECEIVE_FAILURE: {
      return { ...state, status: Actions.REQUEST_STATUS.ERROR, nodes: action.nodes, links: action.links }
    }
    }
  }

  switch (action.type){
  case '@@INIT':{
    return initialState
  }
  case Actions.TOPOLOGY_FILTERS_REQUEST: {
    return {...state,
      filtersStatus: Actions.REQUEST_STATUS.IN_PROGRESS,
    }
  }
  case Actions.TOPOLOGY_FILTERS_RECEIVE_ERROR: {
    return {...state,
      filtersStatus: Actions.REQUEST_STATUS.ERROR,
      err: action.err,
    }
  }
  case Actions.TOPOLOGY_RESTORE_SAVED_FILTERS: {
    const {activeFilters, otherTypeFilters} = getFilterState(state)
    return {...state, activeFilters, otherTypeFilters, savingFilters: true}

  }
  case Actions.TOPOLOGY_NAME_SEARCH: {
    const { searchName } = action
    return {...state, searchName }
  }
  case Actions.TOPOLOGY_FILTERS_RECEIVE_SUCCESS: {
    // The 'clusters' filter is different from other filters.
    // Here we are building the filter options using the cluster labels. When a filter is
    // is selected, we have to use the clusters associated with the label (filterValues).
    const clusterFilters = []
    action.clusters.forEach(({metadata:c}) => {
      clusterFilters.push({
        label: `name: ${c.name}`, //FIXME: NLS. Labels received from the API aren't translated either.
        filterValues: [c.name],
      })
      Object.keys(c.labels).forEach(labelKey => {
        const existingLabel = clusterFilters.find(l => l.label === `${labelKey}: ${c.labels[labelKey]}`)
        if(existingLabel) {
          existingLabel.filterValues.push(c.name)
        }
        else {
          clusterFilters.push({
            label: `${labelKey}: ${c.labels[labelKey]}`,
            filterValues: [c.name],
          })
        }
      })
    })

    // consolidate misc types into "other" type
    const otherTypeFilters = []
    let types =  action.types.map(i => ({label: i }))
    if (types.length>0) {
      types = types.filter(({label})=>{
        if (defaultTypeFilters.has(label)) {
          return true
        } else {
          otherTypeFilters.push(label)
          return false
        }
      })
      types.push({label:'other'})
    }

    const filters = {
      clusters: clusterFilters,
      labels: action.labels.map(l => ({label: `${l.name}: ${l.value}`, name: l.name, value: l.value })),
      namespaces: lodash.uniqBy(action.namespaces, 'metadata.name').map(n => ({ label: n.metadata.name})),
      types,
    }
    return {...state,
      availableFilters: filters,
      otherTypeFilters,
      filtersStatus: Actions.REQUEST_STATUS.DONE,
    }
  }
  case Actions.TOPOLOGY_FILTERS_UPDATE: {
    const activeFilters = {...state.activeFilters} || {}
    activeFilters[action.filterType] = action.filters
    if (state.savingFilters) {
      const {namespace, name} = action
      const cookieKey = namespace ? `${HCM_TOPOLOGY_FILTER_COOKIE}--${namespace}--${name}` : `${HCM_TOPOLOGY_FILTER_COOKIE}`
      localStorage.setItem(cookieKey, JSON.stringify({filterState: {activeFilters, otherTypeFilters: state.otherTypeFilters}}))
    }
    return {...state, activeFilters}
  }
  case Actions.TOPOLOGY_REQUIRED_FILTERS_RECEIVE_SUCCESS: {
    const {item, staticResourceData: {getRequiredTopologyFilters}} = action
    const {metadata: {namespace, name}} = item
    const {activeFilters, otherTypeFilters} = getFilterState(state, namespace, name)
    const requiredFilters = getRequiredTopologyFilters(item)
    activeFilters.label = requiredFilters.label
    return {...state, activeFilters, otherTypeFilters, requiredFilters, savingFilters: true}
  }
  default:
    return { ...state }
  }
}

const getFilterState = (state = initialState, namespace, name) => {
  const cookieKey = namespace ? `${HCM_TOPOLOGY_FILTER_COOKIE}--${namespace}--${name}` : `${HCM_TOPOLOGY_FILTER_COOKIE}`
  let otherTypeFilters = []
  let activeFilters = {...state.activeFilters} || {}
  const savedActiveFilters = localStorage.getItem(cookieKey)
  if (savedActiveFilters) {
    try {
      const savedState = JSON.parse(savedActiveFilters)
      if (savedState.filterState) {
        ({activeFilters, otherTypeFilters} = savedState.filterState)
      } else {
        // legacy
        activeFilters = savedState
      }
    } catch (e) {
      //
    }
  }
  return {activeFilters, otherTypeFilters}
}


