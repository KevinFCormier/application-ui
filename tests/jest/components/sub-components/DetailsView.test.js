/*******************************************************************************
 * Licensed Materials - Property of IBM
 * (c) Copyright IBM Corporation 2018. All Rights Reserved.
 *
 * Note to U.S. Government Users Restricted Rights:
 * Use, duplication or disclosure restricted by GSA ADP Schedule
 * Contract with IBM Corp.
 *******************************************************************************/
'use strict'

import React from 'react'
import renderer from 'react-test-renderer'

import { DetailsView } from '../../../../src-web/components/topology/DetailsView'

describe('DetailsView component 1', () => {
  it('renders as expected', () => {
    const details = [{
      type: 'label',
      labelKey: 'label-key',
      value: 'value',
    }
    ]
    const component = renderer.create(
      <DetailsView
        context={{test: 'US-en'}}
        details={details}
        onClose={jest.fn()}
        resourceType={'test'}
        title={'test'}
      />
    )
    expect(component.toJSON()).toMatchSnapshot()
  })
})
