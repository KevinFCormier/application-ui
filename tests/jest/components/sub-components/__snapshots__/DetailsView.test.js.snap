// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DetailsView component 1 renders as expected 1`] = `
<section
  className="topologyDetails test"
>
  <h3
    className="detailsHeader"
  >
    <svg
      height="48px"
      viewBox="0 0 48 48"
      width="48px"
    >
      <circle
        className="test detailsIcon"
        cx="24"
        cy="24"
        r="20"
      />
      <circle
        className="test"
        cx="24"
        cy="24"
        r="4"
      />
    </svg>
    <span
      className="titleText"
    >
      test
    </span>
    <svg
      alt="Close topology details view"
      aria-label="Close topology details view"
      className="closeIcon"
      fill={undefined}
      fillRule="evenodd"
      height="10"
      name="icon--close"
      onClick={[MockFunction]}
      role="img"
      style={undefined}
      viewBox="0 0 10 10"
      width="10"
    >
      <title>
        Close topology details view
      </title>
      <path
        d="M9.8 8.6L8.4 10 5 6.4 1.4 10 0 8.6 3.6 5 .1 1.4 1.5 0 5 3.6 8.6 0 10 1.4 6.4 5z"
      />
    </svg>
  </h3>
  <hr />
  <div
    className="sectionContent"
  >
    <span
      className="label"
    >
      : 
    </span>
    <span
      className="value"
    >
      value
    </span>
  </div>
</section>
`;
